source/s3_checksums.c

```diff
@@ -0,0 +1,224 @@
+    return aws_hash_update(hash, to_checksum);
+}
+
+int s3_hash_finalize(struct aws_checksum *checksum, struct aws_byte_buf *output, size_t truncate_to) {
+    struct aws_hash *hash = (struct aws_hash *)checksum->impl;
+    checksum->good = false;
+    return aws_hash_finalize(hash, output, truncate_to);
+}
+
+struct aws_checksum *aws_hash_new(struct aws_allocator *allocator, aws_hash_new_fn hash_fn) {
+    struct aws_checksum *checksum = aws_mem_acquire(allocator, sizeof(struct aws_checksum));
+    struct aws_checksum_vtable *vtable = aws_mem_acquire(allocator, sizeof(struct aws_checksum_vtable));
+    struct aws_hash *hash = hash_fn(allocator);
+    checksum->impl = (void *)hash;
+    checksum->allocator = allocator;
+    vtable->update = s3_hash_update;
+    vtable->finalize = s3_hash_finalize;
+    vtable->destroy = s3_hash_destroy;
+    checksum->vtable = vtable;
+    checksum->good = true;
+    checksum->digest_size = hash->digest_size;
+    return checksum;
+}
+
+struct aws_checksum *aws_sha256_checksum_new(struct aws_allocator *allocator) {
+    return aws_hash_new(allocator, aws_sha256_new);
+}
+
+struct aws_checksum *aws_sha1_checksum_new(struct aws_allocator *allocator) {
+    return aws_hash_new(allocator, aws_sha1_new);
+}
+
+struct aws_checksum *aws_md5_checksum_new(struct aws_allocator *allocator) {
+    return aws_hash_new(allocator, aws_md5_new);
+}
+
+struct aws_checksum *aws_checksum_new(struct aws_allocator *allocator, enum aws_s3_checksum_algorithm algorithm) {
+    switch (algorithm) {
+        case AWS_CRC32C:
+            return aws_crc32c_checksum_new(allocator);
+        case AWS_CRC32:
+            return aws_crc32_checksum_new(allocator);
+        case AWS_SHA1:
+            return aws_sha1_checksum_new(allocator);
+        case AWS_SHA256:
+            return aws_sha256_checksum_new(allocator);
+        case AWS_MD5:
+            return aws_md5_checksum_new(allocator);
+        default:
+            return NULL;
+    }
+}
+
+void aws_crc_destroy(struct aws_checksum *checksum) {
+    aws_mem_release(checksum->allocator, checksum->vtable);
+    aws_mem_release(checksum->allocator, checksum);
+}
+
+uint32_t aws_crc32_common(
+    uint32_t previous,
+    const struct aws_byte_cursor *buf,
+    uint32_t (*checksum_fn)(const uint8_t *, int, uint32_t)) {
+
+    size_t length = buf->len;
+    uint8_t *buffer = buf->ptr;
+    uint32_t val = previous;
+    while (length > INT_MAX) {
+        val = checksum_fn(buffer, INT_MAX, val);
+        buffer += (size_t)INT_MAX;
+        length -= (size_t)INT_MAX;
+    }
+    return checksum_fn(buffer, (int)length, val);
+}
+
+int aws_crc32_checksum_update(struct aws_checksum *checksum, const struct aws_byte_cursor *buf) {
+    if (!checksum->good) {
+        return aws_raise_error(AWS_ERROR_INVALID_STATE);
+    }
+    checksum->impl = (void *)(uintptr_t)aws_crc32_common((uint32_t)(uintptr_t)checksum->impl, buf, aws_checksums_crc32);
```

Comment:
This is deeply confusing. It appears that the impl for 32 bit checksums is the running checksum.  Looking at the tortuous casts, I would just make a simple impl that contains the actual value.